Fix publish/delete of unindexed datasets

dbutenhof opened the pull request:
PBENCH-1247

The bulk Elasticsearch API operations (update and delete) have never had clean handling of un-indexed datasets. The functional tests, for example, expected "publish" of the archive-only datasets to fail. This was partly because I did too much "common stuff" inside the bulk action generator. Additionally, a dataset where indexing fails becomes somewhat "stranded" as we expect it to be indexed, but it isn't.

This PR refactors the checks in order to move some common operations out of the action generators with additional checks for eligibility and cleanup. We don't prevent publish, for example, due to lack of an index map if the dataset is archive-only, or if indexing failed, and we'll still do the necessary finalization.

webbnh reviewed (commented):
Looks good, there are just a few editorial things which you should consider fixing up.  (And, I have a couple of nits and suggestions.)  And, I have a question about the no-map test cases.

I restarted your test run -- it failed for what I hope is a external transient.  I'll cycle back after it passes.

dbutenhof commented on lib/pbench/server/api/resources/query_apis/__init__.py:
```diff
@@ -568,9 +569,50 @@ def __init__(
+        ignore the requirement (which is to be sure we don't strand the
+        Elasticsearch documents).
+
+        There are several factors that might tell use not to expect an index
+        map to appear:
```
Ha; when I started the comment I actually thought it was going to be three, but the third one evaporated and didn't take the word with it.

dbutenhof commented on lib/pbench/server/api/resources/query_apis/__init__.py:
```diff
@@ -760,19 +801,34 @@ def _bulk_dispatch(
+            sync.update(dataset=dataset, state=OperationState.WORKING)
+            context["sync"] = sync
+        except Exception as e:
+            current_app.logger.warning(
+                "{} {} setting {} operational state: '{}'",
+                component,
+                dataset,
+                OperationState.WORKING.name,
+                e,
             )
```
It's a `warning` and includes the exception text, so redundant additional verbosity seemed unnecessary. I'm not issuing a warning because everything worked ...

dbutenhof commented on lib/pbench/server/api/resources/query_apis/__init__.py:
```diff
@@ -819,12 +874,15 @@ def _bulk_dispatch(
 
         # Let the subclass complete the operation
         try:
+            current_app.logger.info(f"{dataset.name} is completing")
```
Dang; I thought I caught all of the debugging messages I'd added. 😦

dbutenhof commented on lib/pbench/server/api/resources/query_apis/dataset.py:
```diff
@@ -79,9 +74,31 @@ def __init__(self, config: PbenchServerConfig):
 
+    def prepare(self, params: ApiParams, dataset: Dataset, context: ApiContext):
+        """Prepare for the bulk operation
+
+        This is an empty abstract method that can be overridden by a subclass.
```
It's not technically "stale" ... it was copied from the abstract signature and failed to edit itself as specified. You just can't trust docstrings these days!

dbutenhof commented on lib/pbench/server/api/resources/query_apis/dataset.py:
```diff
@@ -79,9 +74,31 @@ def __init__(self, config: PbenchServerConfig):
+            params: Type-normalized client request body JSON
+            dataset: The associated Dataset object
+            context: The operation's ApiContext
+        """
+        if context["attributes"].action == "update":
```
I'm still naturally biased against multiple return points even though I've deliberately used them more than I once did...

dbutenhof commented on lib/pbench/server/api/resources/query_apis/dataset.py:
```diff
@@ -100,26 +116,13 @@ def generate_actions(
-                if not authorized_user.is_admin():
-                    raise UnauthorizedAdminAccess(authorized_user, OperationCode.UPDATE)
-                context["owner"] = es_doc["owner"] = owner
+            if "access" in context:
+                es_doc["access"] = context["access"]
+            if "owner" in context:
+                es_doc["owner"] = context["owner"]
```
While the set notation is cute and compact, this kinda feels like just "for the sake of being tricky"; and, technically, if I were to let myself be tempted, I think the `fields` local is unnecessary indirection ...

webbnh replied:
> this kinda feels like just "for the sake of being tricky"

It is indeed very close to that (principally because there are only two fields; if there were six, it would be a rather different story...).

However, the definition of `fields` (which could be given a substantially more descriptive name...) could be placed in some common or global location, such that adding another field could be done centrally and wouldn't require adding another `if` statement here just to copy it...the utility of which prompted me to mention it.  (And, I can imagine other uses and code-generalizations based on it....)

For just two fields, it is likely overkill...the question is, if we were to add another field, would we then switch to make the code data-driven?  What about the field after that??  (I worked on a project where they had code like this, and there were about 25 or 30 fields...and, because they had added them one or a few at a time, they never got around to changing the cascading `if` statements into a table....)

dbutenhof commented on lib/pbench/test/unit/server/query_apis/test_datasets_delete.py:
```diff
@@ -213,6 +220,53 @@ def test_no_index(
+        Elasticsearch operations.
+        """
+        monkeypatch.setattr(Metadata, "getvalue", lambda d, k: True)
+        self.fake_elastic(monkeypatch, {}, True)
+
```
Sigh; yeah, I'd meant to look at the existing mock to see if it marked whether it was called and check that (or add it), but I ran out of time last night and lost my context block this morning.

webbnh commented:
The tests look good.  I'm assuming based on your comments that you have an update coming, so I'll wait on approving for that.

dbutenhof commented:
> The tests look good.

There you go, jinxing it again ...

I found some holes, fixed and cleaned up, and found some tests weren't doing what they were supposed to do.

> I'm assuming based on your comments that you have an update coming, so I'll wait on approving for that.

Yeah.

webbnh commented on lib/pbench/server/api/resources/query_apis/__init__.py:
```diff
@@ -577,8 +575,8 @@ def expect_index(self, dataset: Dataset) -> bool:
 
-        There are several factors that might tell use not to expect an index
-        map to appear:
+        We don't expect an index map (and therefore can proceed with a bulk
+        index operation) if:
```
The concept of doing a bulk index operation is beyond the scope of this function.  I would replace the parenthetical with `(and therefore we return 'False')`.

webbnh commented on lib/pbench/test/unit/server/query_apis/test_datasets_delete.py:
```diff
@@ -102,6 +103,40 @@ def fake_bulk(
 
         monkeypatch.setattr("elasticsearch.helpers.streaming_bulk", fake_bulk)
 
+    def unelastic(self, monkeypatch):
```
Should that be "inelastic"?  😁

webbnh commented on lib/pbench/test/unit/server/query_apis/test_datasets_delete.py:
```diff
@@ -102,6 +103,40 @@ def fake_bulk(
+
+            Raises:
+                Exception
+            """
+            raise Exception("We aren't allowed to be here")
```
Insert here the usual caveat about the fact that raising exceptions inside mocks is an unreliable way to cause a test (or anything else) to fail....  (I _would_ recommend using [`pytest.fail()`](https://docs.pytest.org/en/7.1.x/reference/reference.html#id9) instead of `raise`, but, apparently, [it just raises an exception](https://docs.pytest.org/en/7.1.x/_modules/_pytest/outcomes.html#fail).  😝)

webbnh commented on lib/pbench/test/unit/server/query_apis/test_datasets_delete.py:
```diff
@@ -102,6 +103,40 @@ def fake_bulk(
+            raise_on_exception: bool = True,
+        ):
+            """
+            Helper function to mock the Elasticsearch helper streaming_bulk API,
+            which will validate the input actions and generate expected responses.
```
This docstring looks ~stale~ unreliable.

webbnh commented on lib/pbench/test/unit/server/query_apis/test_datasets_update.py:
```diff
@@ -322,7 +358,7 @@ def test_invalid_owner_params(
             f"{server_config.rest_uri}/datasets/{ds.resource_id}",
             headers={"authorization": f"Bearer {pbench_admin_token}"},
-            query_string={"owner": str("invalid_owner")},
+            query_string={"owner": "invalid_owner"},
```
🎉

webbnh reviewed (approved):
Looks good to me, unless you want to update those two docstrings....  (I guess my other comments aren't actionable.)

webbnh reviewed (approved):
👍
